essentials-framework/Essentials Core/PepperDashEssentialsBase/Queues/GenericQueue.cs

```diff
@@ -0,0 +1,156 @@
+﻿using System;
+using Crestron.SimplSharp;
+using Crestron.SimplSharpPro.CrestronThread;
+using PepperDash.Core;
+
+namespace PepperDash_Essentials_Core.Queues
+{
+    /// <summary>
+    /// Threadsafe processing of queued items with pacing if required
+    /// </summary>
+    /// <typeparam name="T">Type of item to be processed</typeparam>
+    public class GenericQueue<T> : IQueue<T> where T : class
```

Comment:
I'm thinking that we may want to create an IQueueMessage interface, then restrict messages here to implement that interface. That makes it easier to create the action to process the queue too. Thoughts?

Reply:
Sure something like the below pseudo code-

```
public interface IQueueMessage
{
   void Dispatch()
}
```

and then the queue would just invoke the dispatch method?

Reply:
Then you could have something like 

```
public class MessageDispatch : IQueueMessage
{
  private readonly IBasicCommunication _coms;
  private readonly string _message;

  public MessageDispatch(IBasicCommunication coms, string message)
  {
    _coms = coms
   _message = message;
  }

public void Dispatch()
{
   _coms.SendText(_message);
}
```

Reply:
Exactly.

Reply:
Sounds good, let me work it up

Reply:
Then on the `GenericQueue` class you can do `public class GenericQueue<T> where T:IQueueMessage`